### PR TITLE
8329096: G1: Change the type of G1BlockOffsetTable::_offset_base to uint8_t*

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

This patch changes the type of `G1BlockOffsetTable::_offset_base` to `uint8_t*`, just like `SerialBlockOffsetSharedArray::_offset_base` and `ObjectStartArray::_offset_base`. The tests `make test-tier1_gc` passed locally.

Thanks for taking the time to review.

Best Regards,
-- Guoxiong

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329096](https://bugs.openjdk.org/browse/JDK-8329096): G1: Change the type of G1BlockOffsetTable::_offset_base to uint8_t* (**Enhancement** - P5)

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [ccae364a](https://git.openjdk.org/jdk/pull/18502/files/ccae364aebef7144486dc9eec4990cbb8798c205)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [ccae364a](https://git.openjdk.org/jdk/pull/18502/files/ccae364aebef7144486dc9eec4990cbb8798c205)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18502/head:pull/18502` \
`$ git checkout pull/18502`

Update a local copy of the PR: \
`$ git checkout pull/18502` \
`$ git pull https://git.openjdk.org/jdk.git pull/18502/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18502`

View PR using the GUI difftool: \
`$ git pr show -t 18502`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18502.diff">https://git.openjdk.org/jdk/pull/18502.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18502#issuecomment-2021566265)

### CODE REVIEW

**Review by albertnetymk — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.cpp [hunk 1]
```diff
@@ -102,18 +102,18 @@ G1BlockOffsetTablePart::G1BlockOffsetTablePart(G1BlockOffsetTable* array, HeapRe
...
-void G1BlockOffsetTablePart::set_remainder_to_point_to_start_incl(u_char* start_card, u_char* end_card) {
+void G1BlockOffsetTablePart::set_remainder_to_point_to_start_incl(uint8_t* start_card, uint8_t* end_card) {
   assert(start_card <= end_card, "precondition");
   assert(start_card > _bot->entry_for_addr(_hr->bottom()), "Cannot be first card");
   assert(_bot->offset_array(start_card-1) < CardTable::card_size_in_words(),
          "Offset card has an unexpected value");
-  u_char* start_card_for_region = start_card;
-  u_char offset = max_jubyte;
+  uint8_t* start_card_for_region = start_card;
+  uint8_t offset = max_jubyte;
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1BlockOffsetTable.cpp:111]: Pre-existing: That should probably be `UINT8_MAX` from stdint/cstdint.h (should already be included).

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.inline.hpp [hunk 2]
```diff
@@ -58,45 +58,44 @@ inline HeapWord* G1BlockOffsetTablePart::block_start_reaching_into_card(const vo
...
-    (const_cast<u_char*> (left), offset, num_cards);
+  memset_with_concurrent_readers(left, offset, num_cards);
 }
 
-inline u_char* G1BlockOffsetTable::entry_for_addr(const void* const p) const {
+inline uint8_t* G1BlockOffsetTable::entry_for_addr(const void* const p) const {
   assert(_reserved.contains(p),
          "out of bounds access to block offset table");
-  u_char* result = const_cast<u_char *>(&_offset_base[uintptr_t(p) >> CardTable::card_shift()]);
+  uint8_t* result = const_cast<uint8_t *>(&_offset_base[uintptr_t(p) >> CardTable::card_shift()]);
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1BlockOffsetTable.inline.hpp:93]: Pre-existing:
```suggestion
  uint8_t* result = const_cast<uint8_t*>(&_offset_base[uintptr_t(p) >> CardTable::card_shift()]);
```

**Review by tschatzl — APPROVED** [review]
lgtm. Maybe fix the minor pre-existing issues.

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.cpp (hunk 1 above)
Comment (lgxbslgx) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.cpp]: Fixed.

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.inline.hpp (hunk 2 above)
Comment (lgxbslgx) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.inline.hpp]: Fixed. And I polished the same problem in `g1BlockOffsetTable.cpp` too.

**Comment by lgxbslgx** [comment]
@albertnetymk @tschatzl Thanks for the review.

/integrate

**Pull request closed without merging**